Improves the UX of menu management in the navigation block

draganescu opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Closes #42257, #42602,  #38169

## Why?
Simplifying the UX of the navigation block. The crowding of the block toolbar with all kinds of disconnected actions does not appear as a good path. 

## How?
A few tweaks:

- [x] move the items in the "select menu" block control to the block's inspector as follows:
  - [x] the list of menus as a dropdown in an inspector panel
  - [X] the manage menus action as a link button below the selector

**Optional:**
- [ ] add a list view representing the tree of InnerBlocks of the navigation block to the block's inspector
  - [ ] allow adding new menu items from the list view representation
  - [ ] disable inner block selection from the list view
  - [ ] cull some of the useless list view item options (e.g. make template part)


**To do**

- [x] There has to be a chevron link in the dropdown toggle
- [x] The label of the currently selected menu does not update on change
- [x] The label of the currently selected menu does not update if menu is renamed from advanced
- [x] Use a visually hidden label to explain this is a menu switcher
- [x] Default to a button if there are no block menus and no classic menus
- [x] Rename "Loading options ..." to "Loading ..."
- [x] Rename "Select another menu" to "Select menu"
- [x] Removed [disabled](https://github.com/WordPress/gutenberg/pull/42987#discussion_r962116873) and implement `isBusy`
- [ ] Update the navigation tests according to the new UI of the block

## Testing Instructions
1. Check out this PR
2. Add a navigation block
3. Behold the amazing UX
4. Test that:

- the menu can be switched to another menu
- importing classic menus work
- new menus can be created
- renaming menus in advanced is immediately reflected in the control
- keyboard navigation works as expected
- screen readers announce this control properly

## Screenshots or screencast <!-- if applicable -->

https://user-images.githubusercontent.com/107534/187418826-c999c81d-64f2-42d5-97c1-063b149ce367.mp4

draganescu committed:
adds a list view with the navigation block inner blocks to the block's inspector

draganescu committed:
adds a first toolpanel item

draganescu commented:
Status:
- shows the menu tree in a list view
- I have some issues with making the tool panel menu work with menu actions:
  - for example to make it show the more icon instead of a plus icon
  - also to include action buttons instead of toggles for tool panel item

I may be using the wrong component.

draganescu committed:
stub toolpanel item

talldan commented:
Reminds me of https://github.com/WordPress/gutenberg/pull/18202 😄 

edit: just for reference, that was removed in #23022 (issue #20846, also see #19691). Something not captured in that issue, but I remember was an issue is that when there are a lot of menu items, the sidebar content can be quite lengthy. Other block tools get pushed out of view and a lot of scrolling is required.

draganescu committed:
moves the menu selector block controls to the inspector

draganescu committed:
fixes label update bug

draganescu committed:
labels navigation menu selector if no menus or deleted menu, fixes label when importing classic menu

draganescu committed:
adds the chevron

draganescu commented:
Maybe for ease of review I'll create a separate PR adding the list view back. It also seems that the inspector of this block is getting extremely cluttered.

draganescu committed:
only show no menus label if no block menus exist, disregard classic menus

draganescu commented on packages/block-library/src/navigation/edit/navigation-menu-selector.js:
```diff
@@ -77,10 +105,11 @@ function NavigationMenuSelector(
 	}
```
I am not sure if this is needed anymore, in the sense that the control should no longer hide itself but the block should hide the whole panel in this case.

MaggieCabrera commented:
This is a much better UX, I agree. It makes more sense to me to handle this in the sidebar. It's working as intended for me

MaggieCabrera reviewed (approved):
LGTM!

draganescu commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -631,8 +630,16 @@ function Navigation( {
 						/>
-					</ToolbarGroup>
-				</BlockControls>
+						<Button
```
I should check for manage menus permission here!

draganescu committed:
adds check for permissions on the manage menus link; removes rogue showManageActions

draganescu commented:
I found a `showManageActions` in the Navigation block edit component that was undefined but passed to the menu selector, defaulting to true. I removed it.

getdave commented on packages/block-library/src/navigation/edit/navigation-menu-selector.js:
```diff
@@ -24,7 +24,6 @@ function NavigationMenuSelector(
 		onSelectNavigationMenu,
 		onSelectClassicMenu,
 		onCreateNew,
-		showManageActions = false,
```
Honestly this depends entirely on the new UX you've implemented. I believe it was originally there in order to allow us to determine whether or not the "Manage menus" options were displayed based on the context in which the NavigationMenuSelector was being shown.

If those different context have now been removed then this requirement is no longer valid and you can remove it.

getdave reviewed (commented):
Thanks for your work here. I believe the goal of reducing the UX "noise" currently associated with the block is a good one.

I noticed a few issues that I think we should look to resolve prior merging this. 

<details>
  <summary>The menu selector is lacking a border.</summary>

<img width="342" alt="Screen Shot 2022-08-26 at 10 09 18" src="https://user-images.githubusercontent.com/444434/186869887-1c09df37-42b8-4d13-8953-27003e8977fa.png">

</details>

<details>
  <summary>Whilst loading the Menus section is empty and then changes to "Create new menu" even though menus are ultimately going to load.</summary>

![Screen Capture on 2022-08-26 at 10-10-37](https://user-images.githubusercontent.com/444434/186870193-4e53d1d4-d643-4a16-a85b-c8578395181e.gif)

</details>


<details>
  <summary>The `Manage menus` link seems to disappear and reappear each time you select a Menu.
</summary>

[![Screen Capture on 2022-08-26 at 10-10-37](https://user-images.githubusercontent.com/444434/186870193-4e53d1d4-d643-4a16-a85b-c8578395181e.gif)
](https://user-images.githubusercontent.com/444434/186870455-a91e8462-4626-4ab1-80ae-c61270125058.mp4)

</details>

<details>
  <summary>The UX when converting a classic menu doesn't provide good feedback. The Menus panel still shows the old menu and then suddenly updates. This leaves me able to change menus or even create new ones whilst the conversion process is happening. We need to improve the guards and gates using the data exposed by the conversion process utility hook.
</summary>

https://user-images.githubusercontent.com/444434/186870754-39a25380-dbf6-48c1-a9a2-85aa58540d25.mp4

</details>


<details>
  <summary>
Similar problem with `Create Menu` action. It might be good to merge [#43529](https://github.com/WordPress/gutenberg/pull/43529) (open with no reviews 😢 ) and _then_ look to take the learnings forward into this PR.
</summary>

https://user-images.githubusercontent.com/444434/186871098-99211840-e807-4a56-8c54-f250bb6a88b8.mp4

</details>

---

Overall this is heading in a good direction. I think we need to do a bit more work to tidy up the UX to ensure we're providing solid feedback to the user.

One thing I would recommend trying that helped me was to test this using a slow internet connect as that quickly uncovers issues that can be masked by using a fast connection. I hope that helps?

draganescu commented:
Great review @getdave ! 😬 all these behaviors were exposed by the moving of the control as the panel in the inspector does not disappear but the toolbar section does and the effect is more jarring. Looking into it.

To do:

- [x] Investigate the lack of border in certain browsers
- I can't repro this one in FF or Chromium
- [x] Optimistically update the label of the selector when converting classic menus
- [x] Bring over the update in https://github.com/WordPress/gutenberg/pull/43529
- [x] Stop hiding the Manage menus link since the perms don't change between menu data fetching
- [x] Handle the whole inspector panel when menu data is loading, or use an accurate label on the selector and disable it

priethor commented:
This PR will likely resolve/invalidate #38169 as well

talldan commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -772,39 +791,42 @@ function Navigation( {
+							clientId={ clientId }
+							onSelectNavigationMenu={ ( menuId ) => {
+								handleUpdateMenu( menuId );
+								setShouldFocusNavigationSelector( true );
```
The code to keep focus on the dropdown doesn't seem to work in this PR. In my testing the `navigationSelectorRef` is `undefined`, but not sure why as the code is much the same.

draganescu replied:
Attempted a fix in [0ede795](https://github.com/WordPress/gutenberg/pull/42987/commits/0ede795475b8b7a5d11055fe813e43226a22fa65)

talldan replied:
It seems to work the first time switching menus, but after that it doesn't work 🤔 

Here are the steps to repro:
(Using the keyboard)
1. Open the navigation menu selector
2. Switch to another menu
3. Expected: the menu should stay open with the menu item still selected
4. Repeat steps 1-3

draganescu replied:
I think I solved that, may need a rebuild?

cbravobernal replied:
It's working on my side 🎉

talldan commented on packages/block-library/src/navigation/edit/navigation-menu-selector.js:
```diff
@@ -64,23 +83,32 @@ function NavigationMenuSelector(
 
+	useEffect( () => {
+		if ( ! hasNavigationMenus ) {
+			setSelectorLabel( __( 'No menus. Create one?' ) );
```
This looks a bit unusual in practice:
<img width="270" alt="Screen Shot 2022-08-29 at 3 56 02 pm" src="https://user-images.githubusercontent.com/677833/187133513-4b74649f-73a9-419e-a243-3aa10c04424a.png">

Perhaps there needs to be some thought about what happens if a menu is missing (or deleted) in the context of this change. IMO, it should probably just fallback to whatever the default menu is (e.g. Page List, or another existing menu).

draganescu replied:
The only problem is what happens when we display the page list fallback. If the navigation block falls back to some menu this UI change is fine. The problem is if the block displays a page list, what then? I chose this direct way to telling the user that even if they have inserted a navigation block, which displays something, they  have not "created" a menu yet.

talldan replied:
Often it shows this when there's a ref, but no navigation entity:
<img width="828" alt="Screen Shot 2022-08-30 at 11 10 43 am" src="https://user-images.githubusercontent.com/677833/187325834-8cab82cd-cdb5-479c-95ce-72ed1b124413.png">

And I think there's an opportunity to simplify this and just fallback to whatever the default is in this instance.

> I chose this direct way to telling the user that even if they have inserted a navigation block, which displays something, they have not "created" a menu yet.

If there needs to be a CTA to create a menu when falling back to a page list, then perhaps it should just be a button rather than a dropdown with one option? Although I also think it would be fine just to stick with the current behavior where a menu is created when the block is edited.

It might be good to loop in a designer on these questions.

draganescu replied:
Yes, I'll post a recording of how it currently works and loop in design feedback as I am really stretching my opinions out on this one :D

cbravobernal replied:
If there is a classic menu available, then that dropdown would have two items. I'm not a designer, but my humble opinion is that keeping the dropdown will be easier for the user as you will have always the same interaction (both for no menus and menus on the list)

talldan commented on packages/block-library/src/navigation/edit/navigation-menu-selector.js:
```diff
@@ -64,23 +83,32 @@ function NavigationMenuSelector(
 			ref={ forwardedRef }
-			label={ __( 'Select Menu' ) }
-			text={ __( 'Select Menu' ) }
+			label={ selectorLabel }
```
It's noticeable that there's no visible label for this dropdown and now the aria-label is also the menu name. For screen reader users, there's no way to know what this menu does.

Most controls in the inspector seem to have a label. I did wonder if a different component could be used here, like a custom select control, see the typography tools:
<img width="269" alt="Screen Shot 2022-08-29 at 4 37 44 pm" src="https://user-images.githubusercontent.com/677833/187138489-d8543cf2-b78d-4ab5-a55c-1272c42c84bc.png">

draganescu replied:
I'll try to add a label before trying to change from this dropdown to a custom select.

talldan replied:
I don't think you can add a label to a dropdown. It has a button to open it and that's it.

draganescu replied:
Would `aria-describedby` work as a prop for the button and then add a hidden descriptor?

draganescu committed:
Implements handling focus persistence in the Navigation Selector according to #42956
Switching between menus keeps the menu open and focused on the current selection.
Importing classic menus or creating new ones focuses the current navigation block.

Co-authored-by: Daniel Richards <677833+talldan@users.noreply.github.com>

draganescu committed:
It looks like `hasManagePermissions` is not cached between API calls so we refresh this quite often when switching between menus. For a more consistent UI experience we disabe the menu management link untill `hasManagePermissions` is refreshed.

Another option would be to cache this at a component level. But, this could be more accurate?

draganescu committed:
Closes the dropdown when importing or creating new menu as in #43529
Fixes slow connection UX by disabling and enabling the selector and the manage menus link depending on data status.

Co-authored-by: Dave Smith <444434+getdave@users.noreply.github.com>

draganescu commented:
I have updated the PR description with a video with the current behavior. I'd love some design feedback at this point as the UI exemplified in https://github.com/WordPress/gutenberg/issues/42602#issuecomment-1207989169 is covered is covered but raises a few points to address:

- @getdave  and @javierarce is the slow connection behavior of disappearing and appearing items on screen something we address globally? I tried to keep things on screen and disable them during data roundtrips (e.g. the site logo block shows the placeholder then hides it if it finds there is a site logo - on a small enough connection we'd have time to interact with it).
- @javierarce is the behavior I attempted via managing label copy something we find desirable? @talldan raised that "If there needs to be a CTA to create a menu when falling back to a page list, then perhaps it should just be a button rather than a dropdown with one option?" Should we switch to a button?
- @alexstine or @talldan would a hidden descriptor referenced by an `aria-describedby` attribute of the dropdown's button toggle be enough in place of a label? This descriptor would be "Switch menu" as we lost this label because of the current behavior change of the menu selector (issue raised by @talldan)

This PR should not attempt to solve the general behavior of the Navigation block and how entities are loaded for it, as this causes most of the flickering / appearing / disappearing. For example, I think some of the things (like users perms) should be saved until page reload (between renders) - but then this PR is already a soup of refactoring and new functionality, I'd love to not make it worse.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -332,9 +330,11 @@ function Navigation( {
 	const [ detectedOverlayColor, setDetectedOverlayColor ] = useState();
 
-	const handleUpdateMenu = ( menuId ) => {
+	const handleUpdateMenu = ( menuId, focusNavigationBlock = false ) => {
```
Nit: I wonder whether the function signature might be clearer with an options object?

```suggestion
	const handleUpdateMenu = ( menuId, { focusNavigationBlock = false } ) => {
```

That way later in the code where we call this function it will be clearer that the 2nd parameter is for focus.

getdave reviewed (commented):
Thanks for continuing to work on this 👍 

I noticed the following things:

<details>
  <summary>I noticed that the Menu selector in the sidebar doesn't close once a selection has been made. 
</summary>

That is different to `trunk`. Is this intentional?

https://user-images.githubusercontent.com/444434/187438242-122f34d6-4a83-4719-b12e-e4c1d07d2fa5.mp4
</details>

<details>
  <summary>Lower permission users now no longer appear to be able to "read" Navigations created by high permission users. 
</summary>

They should be able to see them but not edit or create. Note also the permissions message is missing when selecting the block.

Update: I checked and this appears to also be broken on trunk 🤦 


https://user-images.githubusercontent.com/444434/187440421-1f01d90a-9512-48d3-84d6-d5c3bc8e088b.mp4



</details>

> is the slow connection behavior of disappearing and appearing items on screen something we address globally? I tried to keep things on screen and disable them during data roundtrips (e.g. the site logo block shows the placeholder then hides it if it finds there is a site logo - on a small enough connection we'd have time to interact with it).

I like what you've done here with keeping elements in the DOM and updating to disabled and displaying appropriate messages. As long as the wider states are communicated by `speak()` and friends then we should be ok.

javierarce commented:
> @getdave and @javierarce is the slow connection behavior of disappearing and appearing items on screen something we address globally? I tried to keep things on screen and disable them during data roundtrips (e.g. the site logo block shows the placeholder then hides it if it finds there is a site logo - on a small enough connection we'd have time to interact with it).

What you've done looks good to me, but I have two comments: I think the link to manage menus shouldn't appear disabled, and instead of "Loading options…" I would just go with "Loading…"

> @javierarce is the behavior I attempted via managing label copy something we find desirable? 

Which behavior do you mean? 


> @talldan raised that "If there needs to be a CTA to create a menu when falling back to a page list, then perhaps it should just be a button rather than a dropdown with one option?" Should we switch to a button?

<details>
<summary>
Yes,  in that case switching to a button would be OK. I think we could even hide the link to manage menus
</summary>

![image](https://user-images.githubusercontent.com/4933/187459783-cafd89f8-796e-4e5d-9f46-b28de9878b46.png)
 
</details>

alexstine commented:
@draganescu The aria-describedby attribute should only be used for descriptions as screen readers treat them differently. You should either define proper button text or add an aria-label. In Voiceover for example, aria-describedby is not exposed by default in most verbosity settings. Voiceover would announce something like this.

> Name of button text/aria-label, item has help tag

The help tag in Voiceover is defined by aria-describedby. This is why it is extremely important to use these attributes correctly.

If possible, why not import VisuallyHidden from wordpress/a11y and define the text you need? At least we all know screen readers read text the same. What is up for debate is how screen readers interpret ARIA attributes.

Thanks.

draganescu commented:
> Which behavior do you mean?

@javierarce I mean that the label of the select changes depending on situation:

- if there are no block manus it says "No menus. Create one?"
- if there is data still fetching "Loading options …"
- if the currently ref'd menu is missing "Select another menu"

>  in that case switching to a button would be OK. 

@javierarce the proposed UI looks great! One thing, if there _are_ classic menus to import we should _not_ default to a button, right?

>  I think the link to manage menus shouldn't appear disabled

@javierarce ideally it shouldn't appear but there is some kind of bug outside the scope of this PR that makes the permission checks repeatedly and this causes the link to appear and disappear if connections are slow. I am disabling/enabling it so that we avoid jittery jumping screen elements. It's not ideal but I think it's better until we solve the original issue of requesting and updating permissions with every menu change/update.

---

@talldan in an earlier comment you mentioned that if the ref'd menu is missing we should go forward with the fallback behavior instead of showing a placeholder. I strongly suggest we keep the placeholder because seeing different content as a result of a user deleting content is very confusing. However @javierarce we can revisit the placeholder and its options in a new PR!

---

> If possible, why not import VisuallyHidden from wordpress/a11y and define the text you need?

@alexstine I'll try this. Apparently `label` works with `button` [just fine](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label) so I can render a visually hidden label.


---

> I noticed that the Menu selector in the sidebar doesn't close once a selection has been made. 

@getdave this is intentional, bringing over the updates from #42956

> Lower permission users now no longer appear to be able to "read" Navigations created by high permission users. 

Let's address this in another PR fixing `trunk` then.

jasmussen commented:
Thanks for working on this. Quick question as it wasn't immediately clear from the video. At the moment, if I have no menus created/saved, it falls back to the page list, like so:

<details>
<img width="1392" alt="Screenshot 2022-08-31 at 09 14 48" src="https://user-images.githubusercontent.com/1204802/187616535-0937a50f-b454-4926-9990-eb6bdd2e5de2.png">
</details>

This is excellent behavior as in most cases that will likely be entirely sufficient. Will this behavior be maintained?

draganescu commented:
@jasmussen yes that behavior is maintained and should be unaffected by this change. This change merely moves the menu selector to the inspector, keeping how the block works the same.

paaljoachim commented:
Thank you for working on this @draganescu Andrei!

Testing.
Using the Gutenberg PR build.
Twenty Twenty Two.
In a new page I added the Nav block.

<details>
<summary>Before/After</summary>

Before:

![Screenshot 2022-09-02 at 16 25 51](https://user-images.githubusercontent.com/5323259/188170950-07bcf1fe-93cf-4ddd-9000-b01df936aae1.jpg)

After:

![Screenshot 2022-09-02 at 16 26 46](https://user-images.githubusercontent.com/5323259/188171000-0f83c965-48d8-453e-a575-d4fbf6e15410.jpg)
</details>

Should the Menu section be seen above the Layout section? As in importance it would be best to first see Menu section and then the Layout section.

Should the menu drop down have a border/outline giving a stronger signal that it is a drop down?

Clicking Manage menus leads one out of the page and into a Navigation Menus screen somewhere in the backend, and not connected with any place. Similar to managing Reusable blocks. Clicking to edit a menu in the Navigation Menus screen one can only edit the name and not the contents of the menu. 
When clicking Manage menu I expected that it would make it easier for me in someway to edit the specific menu contents or that it might take me to the "old" Appearance -> Menus screen.

What else should I check?

My conclusion is that it is nice to add this specific feature into the sidebar. I look forward to having more of the Navigation features in the sidebar. Thanks.

draganescu commented:
Thanks for testing @paaljoachim 👍🏻 

> Should the Menu section be seen above the Layout section? As in importance it would be best to first see Menu section and then the Layout section.

Yes, but for some reason the layout options cannot be overiden yet.

> Should the menu drop down have a border/outline giving a stronger signal that it is a drop down?

What browser were you testing with? I and others cannot repro the lack of border but you're the second to observe it.

> What else should I check?

Updated the PR description with items.

draganescu committed:
Larger commit, adds:

- The label of the currently selected menu updates if menu is renamed from advanced
- Uses a visually hidden label to explain this is a menu switcher
- Defaults to a button if there are no block menus and no classic menus
- Renames "Loading options ..." to "Loading ..."

Co-authored-by: Javier Arce <4933+javierarce@users.noreply.github.com>
Co-authored-by: Alex Stine <13755480+alexstine@users.noreply.github.com>

paaljoachim commented:
Local site using Desktop Server.
Php version 7.3.
Twenty Twenty Two theme.
Gutenberg PR build.

<details>
<summary>
I have now tested with Brave, Firefox and Chrome. None had a grey border around it and looked like this (screenshot from Chrome):
</summary>
![Screenshot 2022-09-02 at 17 45 57](https://user-images.githubusercontent.com/5323259/188191002-853cf04b-cda8-43d7-9862-ae3cea5fabfc.jpg)
</details>

<details>
<summary>
Did an Inspect and noticed the .components-button had border 0. I adjusted it to 1px solid grey.
It now has a grey border around the drop down as does every other button...:) So it is not the place to add it but I did it to experiment.
</summary>

![Screenshot 2022-09-02 at 17 42 55](https://user-images.githubusercontent.com/5323259/188190696-499a8949-fa6c-4cf4-83a2-174c5004c1f6.jpg)
</details>

Btw it took a moment (saw the spinner instead of the Nav block menu for half a second) before the menu showed up.

talldan commented:
> @talldan in an earlier comment you mentioned that if the ref'd menu is missing we should go forward with the fallback behavior instead of showing a placeholder. I strongly suggest we keep the placeholder because seeing different content as a result of a user deleting content is very confusing. However @javierarce we can revisit the placeholder and its options in a new PR!

I think it looks a bit like an error state, so I do think it's worth revisiting. Previously the placeholder was shown, but I guess there is no placeholder now. I see it wasn't introduced by this PR, so don't worry about it right now.

talldan commented on packages/block-library/src/navigation/edit/navigation-menu-selector.js:
```diff
@@ -5,82 +5,172 @@ import {
+			<Button
+				{ ...toggleProps }
+				className="wp-block-navigation__navigation-selector-button--createnew"
+				disabled={ ! enableOptions }
```
`disabled` causes a focus loss, so I'd recommend against using it.

The `Button` has an `isBusy` for loading state. Return early from the `onClick` to prevent multiple clicks.

getdave replied:
> disabled causes a focus loss, so I'd recommend against using it.

TIL

draganescu replied:
I have moved away from `disabled` to `isBusy`. Unfortunately preventing multiple clicks is more complicated.

javierarce commented:
> @javierarce I mean that the label of the select changes depending on situation:

> if there are no block manus it says "No menus. Create one?"
> if there is data still fetching "Loading options …"
> if the currently ref'd menu is missing "Select another menu"
> in that case switching to a button would be OK.

For the sake of consistency and simplicity, we should only have one message "Select menu" (and obviously "Loading..." for the loading state). Having different messages is confusing and could cause problems in other languages. 

So we would have this:

- If we are fetching data: "Loading..."
- If there's a menu selected: the name of the menu
- If there's no menu selected/reference is missing: "Select menu"
- If there are no menus whatsoever: we show a button to create one

> @javierarce the proposed UI looks great! One thing, if there are classic menus to import we should not default to a button, right?

Right, we should show "Select menu" because there are menus (the only difference is that the user will be able to import and select them simultaneously).

> @javierarce ideally it shouldn't appear but there is some kind of bug outside the scope of this PR that makes the permission checks repeatedly and this causes the link to appear and disappear if connections are slow. I am disabling/enabling it so that we avoid jittery jumping screen elements. It's not ideal but I think it's better until we solve the original issue of requesting and updating permissions with every menu change/update.

Gotcha, thanks for the explanation!

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -603,33 +584,44 @@ function Navigation( {
-				</BlockControls>
+						<Button
+							variant="link"
+							disabled={
+								! hasManagePermissions ||
+								! hasResolvedNavigationMenus
+							}
```
Should this using `isBusy` instead of `disabled` now [we know that `disabled` causes focus loss](https://github.com/WordPress/gutenberg/pull/42987#discussion_r962116873)?

draganescu replied:
Yes I added a todo with @talldan 's suggestion

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -671,39 +663,50 @@ function Navigation( {
 								if ( navMenu ) {
-									handleUpdateMenu( navMenu.id );
-									setShouldFocusNavigationSelector( true );
+									handleUpdateMenu( navMenu.id, true );
```
Nit: reading this I'm thinking "what does `true` do"? I have to do to the function definition to find out.

As a result, I really think an options object as second arg with a named parameter would be clearer.

draganescu replied:
Done.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -671,39 +663,50 @@ function Navigation( {
-				</BlockControls>
+						<Button
+							variant="link"
+							disabled={
+								! hasManagePermissions ||
+								! hasResolvedNavigationMenus
+							}
```
Ditto other comments about `disabled` causing focus loss.

draganescu replied:
I think this is OK for the link button as the result is navigating to another page.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -771,39 +772,44 @@ function Navigation( {
+						/>
+						<Button
+							variant="link"
+							disabled={
+								! hasManagePermissions ||
+								! hasResolvedNavigationMenus
+							}
```
Ditto other points about `disabled` causing focus loss.

draganescu replied:
I think this is OK for the link variant as the action is to navigate so focus is lost anyway.

getdave commented on packages/block-library/src/navigation/edit/navigation-menu-selector.js:
```diff
@@ -5,82 +5,172 @@ import {
+			);
+		} else if ( hasResolvedNavigationMenus && currentMenuId === null ) {
+			setSelectorLabel( __( 'Select another menu' ) );
+			setEnableOptions(
+				( canSwitchNavigationMenu || canUserCreateNavigationMenu ) &&
+					hasResolvedNavigationMenus &&
+					( ! isCreatingMenu || createNavigationMenuIsSuccess )
+			);
```
I wonder if some of these could be refactored to well named variables? At the moment one has to parse and understand all these conditionals. Is there a way to make it easy to understand at a glance?

draganescu replied:
Done.

scruffian committed:
move styles to editor.scss as they are only needed for the editor

scruffian committed:
use flex rather than absolute positioning

scruffian committed:
alphabetize CSS properties

getdave reviewed (approved):
I tested this and it's really good. I especially appreciate the work to ensure that the select menus close _immediately_ upon selecting an option that causes an _async_ action to occur.

<details>
<summary>Missing border appears to be coming from `.components-button`.</summary>
<img width="1390" alt="Screen Shot 2022-09-05 at 10 11 17" src="https://user-images.githubusercontent.com/444434/188413348-58bebb79-2f89-456a-bd58-88f9cb5aa821.png">
</details>

<details>
<summary>I noticed that when switching between menus that are similar it's not always obvious that something has happened. We might need a follow up to announce (via a notification) that the menu has changed - especially for users of assistive technology.</summary>

https://user-images.githubusercontent.com/444434/188414195-08804577-9489-4297-9fb2-4202c554efa6.mp4
</details>

scruffian reviewed (approved):
This is a good improvement. I say we merge it and follow up with the necessary changes to the tests in a separate PR.

getdave commented:
<details>
<summary>@draganescu I was testing with TT2. Now I'm on `emptytheme` the border is there 🤔 </summary>
<img width="1412" alt="Screen Shot 2022-09-05 at 10 38 59" src="https://user-images.githubusercontent.com/444434/188419039-8cb051a6-64d2-4037-a7ee-3f95235dee13.png">
</details>

MaggieCabrera commented:
> @draganescu I was testing with TT2. Now I'm on `emptytheme` the border is there 🤔

that's so strange, I was testing this with emptytheme all the time and never got it to show without the border :S

draganescu committed:
Incorporates review feedback:

- replaces disabled with isBusy for toggle working state
- refactores some conditionals for better readability
- adds consistency to the toggle label

Co-authored-by: Daniel Richards <677833+talldan@users.noreply.github.com>
Co-authored-by: Dave Smith <444434+getdave@users.noreply.github.com>
Co-authored-by: Javier Arce <4933+javierarce@users.noreply.github.com>
Co-authored-by: Paal Joachim Romdahl <5323259+paaljoachim@users.noreply.github.com>

javierarce commented:
This is looking great, but there's a small detail we could fix: if the user opens the dropdown and clicks on the already selected menu, the dropdown doesn't get closed. The proper behavior should be to close the dropdown.

draganescu commented:
@getdave it seems that after @scruffian 's commit moving the control styles from `styles.scss` to `editor.scss` the border appears in TT2 as well.

---

@javierarce we specifically implemented in the chances from @talldan that I brought over from [#42956](https://github.com/WordPress/gutenberg/pull/42956) to not close the menu when switching for a11y reasons.

@alexstine if we close the dropdown when selecting the same menu, should we focus the block or the control?

Either way this can be revisited.

---

@talldan Some details on `isBusy` and multiple clicks:

The `DropdownMenu` component does not expose the `onToggle` prop of `Dropdown` and `onToggle` is called before the click handler of the toggle.

I'll leave it as is, as multiple clicks don't damage data, it can only cause buggy UI and will come back with some kind of update for `DropdownMenu`.

This is the second issue I run into with a component in this PR  - the 1st one being with `Popover` because its `onFocusOutside` prop when set skips the call to `onToggle`. I would have needed that to make the arrow of the control reset when the drop down closes on click outside of the control. Instead this is a known bug for now.

draganescu commented:
I will merge this and revisit the following:

- [ ] adding (back) the list view of the navigation block to its inspector
- [ ] fixing `Popover` somehow or the navigation selector via some "hack" to properly toggle the chevron when the popover closes via `onFocusOutside`
- [ ] fixing `DropdownMenu` to be able to prevent it from toggling when the toggle `isBusy`
- [ ] (maybe) closing the menu selector when _the same menu as the one selected_ is clicked.

talldan commented:
> @talldan Some details on isBusy and multiple clicks:
>
> The DropdownMenu component does not expose the onToggle prop of Dropdown and onToggle is called before the click handler of the toggle.

@draganescu I'm not sure I follow. I think it was a normal `Button` that I left the comment on rather than a `DropdownMenu`. The one that appears when there are no menus. <strike>You should be able to return early from onClick to prevent anything bad happening from multiple clicks.</strike>

I guess it may also happen with dropdown menus and that's what you're referring to.

Anyway, I looked a bit deeper and I think I wasn't completely correct with my previous advice. You can use `disabled`, but `__experimentalIsFocusable` needs to be specified to make the `Button` component output `aria-disabled` instead of the normal `disabled` attribute and remain focused. `isBusy` can be used to show a loading indicator, but it depends on whether that fits the design. Here's some pseudocode:
```jsx
<Button isBusy={ isCreatingMenu } disabled={ isCreatingMenu } __experimentalIsFocusable>Create menu</Button>
```

edit: `Button` does also implicitly prevent clicks when configured this way:
https://github.com/WordPress/gutenberg/blob/ec29b9c953f5a1934f5cf24a6010ef1f31004545/packages/components/src/button/index.js#L119-L130

jasmussen commented:
@draganescu It doesn't look like inspector menu editing made it in this round. It still seems valuable, is there any chance we can still land it? 

Edit: from the context it looks like it was an oversight. I've reopened! 👍 👍
